### DISCUSSION TITLE
Changing the mouse cursor in WinUI 3.

### DISCUSSION THREAD

**Discussion by AtomicBlom** [microsoft/WindowsAppSDK#1816] [Q&A] [answered]
With the release of the 1.0 Windows App SDK and WinUI 3 (Congrats team), I've been experimenting with porting our application from UWP to the App SDK.

The one stumbling block I've had getting it to a state where it will compile is changing the mouse cursor. The release notes for the stable channel mention that there are new APIs for changing the cursor bitmap.

From what I've been able to gather, I can set the cursor by using an InputPointerSource, and setting the Cursor property on it.

```csharp
pointer.Cursor = InputSystemCursor.Create(InputSystemCursorShape.IBeam);
```

My issue however is how to actually get a reference to the InputPointerSource, the documentation seems to specifically mention a XamlSwapChain which I didn't think I would need, and Intellisense tells me I can call

```csharp
InputPointerSource.FromAbi(IntPtr)
```

or perhaps

```csharp
new InputPointerSource(IObjectReference)
```

alternatively, I checked the properties of TappedRoutedEventArgs and PointerRoutedEventArgs and I'm able to get a PointerId from the latter and nothing from the former. Ideally, there would be a mechanism to set the cursor for any pointer over my Window (or AppWindow)

Any guidance would be appreciated.

**Comment by castorix** [comment]
A way is to call **set_ProtectedCursor** with System.Reflection,
 but there should be a more direct/simpler way...

A test with Microsoft.UI.Input.InputSystemCursorShape.Hand : 

![set_ProtectedCursor](https://user-images.githubusercontent.com/22345506/142270714-650ad2b4-9096-453d-b57e-cb97f9f77b7f.gif)

**Comment by andrewleader** [comment]
@kevinguo305 can you help answer this question? I think it's in your area path?

**Comment by kevinguo305** [comment, marked as answer]
Getting the CoreCursor from InputPointerSource is not the recommended way. We've added this new API:

```
Microsoft.UI.Xaml.UIElement
{
    protected CoreCursor ProtectedCursor { get; set; }
}
```

and we recommend instead that you subclass the UI element that you want the cursor to be changed in, and add a method that sets a ProtectedCursor property which is part of every UIElement in WinUI 3. 

```
public void ChangeCursor(MUI::InputCursor cursor)
{
	this.ProtectedCursor = cursor;
}
```

Afterwards, in your main Microsoft.UI.Xaml.Controls.Page class, you can call the subclassed UI Element.ChangeCursor(InputSystemCursor.Create(\*insert InputSystemCursorShape enum value\*));

**Reply by AtomicBlom** [reply]
Thanks for your reply. I'll give it a shot and make sure that I'm able to do this and report back once I'm sure it will work in our solution. I have some early lifecycle crashing to deal with before I can verify.

**Reply by AtomicBlom** [reply]
Just wanted to get back to you and confirm that this worked. I was concerned because I needed to subclass Grid, but that caused no problems, a little clunky but this is the best solution.

**Reply by asierpn** [reply]
And how can we do this with TextBlock if it is sealed?

**Reply by nghia-vt** [reply]
Hi @kevinguo305, I did as you recommend, but it throws an exception at `this.ProtectedCursor = cursor;` because the ProtectedCursor property is protected:
> System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.

Below code is my implementation, Am I wrong or missing anything? @AtomicBlom Really appreciate if you can help. Thanks!
```
    public sealed partial class MainPage : Page
    {
        public MainPage()
        {
            this.InitializeComponent();

            var myBtn = new Button2() { Width = 100, Height = 50};
            this.MainPageGrid.Children.Add(myBtn);

            myBtn.ChangeCursor(InputSystemCursor.Create(InputSystemCursorShape.Hand));
        }
    }

    public class Button2 : Button
    {
        public void ChangeCursor(InputCursor cursor)
        {
            this.ProtectedCursor = cursor;
        }
    }
```

**Reply by asierpn** [reply]
@nghia-vt you can not set ProtectedCursor until the Visual Tree is loaded, see https://github.com/MicrosoftDocs/winui-api/issues/112

**Reply by nghia-vt** [reply]
@asierpn Ah yes, I call ChangeCursor in PointerEntered event and it works. Thank you so much!

**Reply by br441** [reply]
@nghia-vt @AtomicBlom What am I doing wrong here? I'm trying to change the cursor for MediaPlayerElement, which is a valid UIElement. It's throwing System.NullReferenceException on the myMPE initialization. 
```
public partial sealed class MainWindow : Window 
{
    public class MainGrid : Grid
    {
        public MPE myMPE = new() { AutoPlay = false, AreTransportControlsEnabled = false };
        public MainGrid()
        {
            Children.Add(myMPE);
            myMPE.ChangeCursor(InputSystemCursor.Create(InputSystemCursorShape.Hand));
        }
    }
    public MainWindow()
    {
        InitializeComponent();
        this.Content = new MainGrid();
    }
}
public class MPE : MediaPlayerElement
{
    public void ChangeCursor(InputCursor cursor)
    {
        ProtectedCursor = cursor;
    }
}
```

**Reply by codendone** [reply]
@br441 Deriving from MediaPlayerElement apparently has a problem: microsoft/microsoft-ui-xaml#8973

**Reply by tipa** [reply]
@codendone @kevinguo305 What is the recommended approach to conditionally set a different cursor in a (Rich)TextBox control? I want to show a hand cursor when the user hovers over a hyperlink. On UWP this worked by setting the `CoreCursor` on the `CoreWindow`. When I now set `ProtectedCursor`, the hand cursor is not visible, only the standard "IBeam" cursor

**Comment by devmikew** [comment]
That's all well and good. However, surely this is a style thing and it should be able to be set in a style, possibly using a visual state change.

**Reply by legistek** [reply]
Wait - so you can't actually change the hover cursor for a control just through XAML? WTH?

**Reply by amercer-fdi** [reply]
> Wait - so you can't actually change the hover cursor for a control just through XAML? WTH?

Yeah it's been 9 months since your comment and nothing has changed. Just trying to get into WinUI3 and I keep running into things like this. What in the world...

**Comment by kevinguo305** [comment]
[benkuhn- redacted by request of the author, but leaving as placeholder because the comment is still useful / relevant]

**Reply by legistek** [reply]
There seems to be an extension property FrameworkElementExtensions.Cursor in the community toolkit but it doesn't seem to do anything. 
After less than a day of attempting to explore porting my WPF app to WinUI it's painfully obvious this thing is not ready for prime time. Not even close. I might as well just make a PWA. I don't understand what Microsoft is thinking making this SDK so unusable on so many levels.

**Comment by marb2000** [comment]
@legistek, for transparency's sake (and to help you to make a better decision for your product), I want to clarify that It's not part of the current WinAppSDK 1.2 engineering plan to expose a XAML property to change the cursor. It's in our backlog, but it was moved out of 1.2.

**Reply by legistek** [reply]
Thanks @marb2000, I appreciate that. Cursor alone isn't a deal breaker - and there's really a lot of great stuff here - but there's just so much missing vs. WPF that porting is going to be nearly a rewrite of all the XAML. I've been waiting for a modern framework that could really succeed WPF for a long time and it's just disappointing that we're not there yet. Add to that that the XAML isn't compatible with MAUI (not your guys' fault, I know, but still a bad Microsoft decision) and it's hard to make the case for initiating this port. Appreciate your transparency though, thank you!

**Reply by amercer-fdi** [reply]
@marb2000, is this on the 1.3 roadmap?

**Comment by devmikew** [comment]
Well you can, but one needs to subclass every control in the universe. You could also use reflection to get to the protected cursor  which shouldn't be done for a production system.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by legistek** [reply]
IMO I would go ahead and use reflection here even in production. It's protected, not internal, so if they changed it it would be a breaking change generally that you could detect at compile time easily enough.

Why was it made protected and not public though? Sadism?

Also I noticed if I change it in the class constructor I get a protected memory error. An actual protected memory error. What's the over/under on how far I can go before getting a BSOD? Makes me nostalgic for MFC.

**Comment by devmikew** [comment]
That's not the only thing that's crazy, eg, look at the Window class, its the top level control in the system and its not even a dependency object.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by marb2000** [comment]
@devmikew If you have to subclass every control in your app to get this behavior, it's clear to me that the framework is not given you the right set of features for your scenario. Although I have seen companies that subclass every control because they add them their _special sauce_. But this should be the expectation, not the rule. 
Also, it's unfortunate that Windows is not a dependency object, so you have to delegate all the common DataContext, Binding, resources, and styles stuff to the main content of the Window object. It's something that we can change until at least version 2.x to keep backward compatibility with all the 1.x releases.

**Comment by devmikew** [comment]
The only way I've been able to do this is by using reflection.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by asierpn** [comment]
Yes, reflection is the only way I've found to do it, I've created an extension method:

    public static void ChangeCursor(this UIElement uiElement, InputCursor cursor)
    {
      Type type = typeof(UIElement);
      type.InvokeMember("ProtectedCursor", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.SetProperty | BindingFlags.Instance, null, uiElement, new object[] { cursor });
    }

And a behavior:

    public static readonly DependencyProperty CursorProperty = DependencyProperty.Register("Cursor", typeof(InputSystemCursorShape), typeof(CursorBehavior), new PropertyMetadata(InputSystemCursorShape.Arrow));
    public InputSystemCursorShape Cursor
    {
      get
      {
        return (InputSystemCursorShape)GetValue(CursorProperty);
      }
      set
      {
        SetValue(CursorProperty, value);
        SetCursor();
      }
    }

    protected override void OnAttached()
    {
      base.OnAttached();
      AssociatedObject.Loaded += AssociatedObject_Loaded;
    }

    protected override void OnDetaching()
    {
      base.OnDetaching();
      AssociatedObject.Loaded -= AssociatedObject_Loaded;
    }

    private bool _loaded;
    private void AssociatedObject_Loaded(object sender, RoutedEventArgs e)
    {
      _loaded = true;
      SetCursor();
    }

    private void SetCursor()
    {
      if (!_loaded)
        return;
      AssociatedObject.ChangeCursor(InputSystemCursor.Create(Cursor));
    }

**Comment by devmikew** [comment]
Good, thanks for your code.

If anyone at MS is watching, please fix this.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by legistek** [comment]
Why not make a `Border` subclass that exposes a cursor dependency property and wrap anything with it whose cursor you need to change through binding?

**Comment by devmikew** [comment]
Yes, that is also possible,  but eventually you will have to subclass all controls. If not you then across all developers. If smells a fair bit .

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by WaGi-Coding** [comment]
Is there anything simpler meanwhile? I love WinUI looks but it makes everything so complicated ^^

**Comment by khoabui1412** [comment]
How can I change cursor to specific image (for ex: pen.cur), I added pen.cur as an embedded resource, and used InputDesktopResourceCursor.Create(uint resourceId), but I have no ideal to find the resourceId of pen.cur.
I am using unpackage app.

**Reply by zkvii** [reply]
[I believe this might help ](https://github.com/microsoft/WindowsAppSDK/discussions/2722#discussioncomment-3662373)

if u get any problems in create .res file, then create a c++ dll project in vs and create .rc file and include ur custom assets then build and copy res to ur winui unpackaged application as the link instructions. 
Hope u can make it, thought the answer might be late.

**Comment by huntran-clgx** [comment]
`ProtectedCursor` now will throw error on 1.4.4

https://github.com/microsoft/microsoft-ui-xaml/issues/9277

**Reply by asierpn** [reply]
The reflection workaround still works on 1.4.4 https://github.com/microsoft/WindowsAppSDK/discussions/1816#discussioncomment-3430802

**Reply by DarranRowe** [reply]
It is not 1.4.4, it is the January 2024 version of C++/WinRT.
You can also query for the IUIElementProtected interface. However, as stated, directly accessing ProtectedCursor is not intended.

**Comment by jeremy-visionaid** [comment]
@kevinguo305 Thanks for clarifying the intended mechanism, but I think as you can see from this thread, having to subclass the control isn't really an acceptable solution to a lot of users/cases. What's the thinking about how 3rd party controls (that might be sealed, etc.) could be controlled in that way, or MAUI for example? Custom cursors were/are so easy in WinForms, and just so unnecessarily hard in WinUI.

**Reply by MikeAnRzMsft** [reply]
Exactly!  it should be in Styles

**Comment by devmikew** [comment]
As well as in code.Sent from my Galaxy
-------- Original message --------From: vmiguelangmsft ***@***.***> Date: 16/12/24  11:19  (GMT+10:00) To: microsoft/WindowsAppSDK ***@***.***> Cc: devmikew ***@***.***>, Mention ***@***.***> Subject: Re: [microsoft/WindowsAppSDK] Changing the mouse cursor in WinUI 3. (Discussion #1816) 
Exactly!  it should be in Styles
